### PR TITLE
bundle parameter now handels NPM modules and globs

### PR DESCRIPTION
**Pull request by pYr0x**
## WIP

first commit for https://github.com/stealjs/steal-tools/issues/507

### CODE REVIEW

File: lib/loader/find_bundle.js [hunk 1]
```diff
@@ -9,30 +14,126 @@ module.exports = findBundles;
...
-function minusJS(name) {
-	var idx = name.indexOf(".js");
-	return idx === -1 ? name : name.substr(0, idx);
+function endsWith(str, suffix) {
+	return str.indexOf(suffix, str.length - suffix.length) !== -1;
 }
+function startsWith(starts, path) {
+	return path.substr(0,starts.length) === starts;
+}
+function isNpm(moduleName){
```
Comment (matthewp) [inline lib/loader/find_bundle.js]: you can require npmUtils and get all of this stuff: `var npmUtils = require("steal/ext/npm-utils");`

Comment (pYr0x) [inline lib/loader/find_bundle.js]: ok. thanks.
i have to ivestigate the problem, why the tests are not passing.

Comment (matthewp) [inline lib/loader/find_bundle.js]: I think major might be failing in general.

Comment (pYr0x) [inline lib/loader/find_bundle.js]: why?

**Comment by pYr0x** [comment]
@matthewp i think i found the problem
https://github.com/stealjs/steal-css/pull/19
can you make a release for steal-css and merge it into steal?

**Comment by matthewp** [comment]
Yep, doing it now, will let you know and you can update this PR.

**Comment by matthewp** [comment]
https://github.com/stealjs/steal/pull/829

**Comment by pYr0x** [comment]
you also add the css.js into steal and make a new rc release?

**Comment by matthewp** [comment]
in the process :)

**Comment by matthewp** [comment]
@pYr0x rc-1 has this. Thanks!

**Comment by pYr0x** [comment]
thanks @matthewp 
the tests are passing. but one is failing
```
1) multi build Source Maps works with configDependencies:
     Error: ENOENT: no such file or directory, open '/home/travis/build/stealjs/steal-tools/test/npm-config-dep/dist/bundles/main.js.map'
```

i think this is releated to https://github.com/stealjs/steal/issues/803 ?

should we merge that PR and open a new PR for stripping the `sourceMapping` ?

**Comment by matthewp** [comment]
I think it would be more maintainable of we just removed the glob pattern and ran it through loader.normalize. That would handle the appName and ~/ conventions and all of that other stuff. I'm afraid this code will break accidentally if we change the npm plugin in the future.

I'm thinking it should be like:

1. Remove `/**/*` (or whatever it is).
2. `loader.normalize(name, loader.main)`
3. Add `/**/*` back on.
4. Do directories.lib stuff (sad that this is needed)
5. Then run through the glob finder.

This would make it so that any normalize hook stuff is covered, like the backslash extension.

**Comment by pYr0x** [comment]
what to you mean by 
> Remove /**/*

we use not only the glob, we also use the [minimatch ](https://www.npmjs.com/package/minimatch) pattern, see the tests.
https://github.com/stealjs/steal-tools/blob/8b128f2f7accd7a4b24741b47882dbd02a41d33d/test/bundle_test.js#L14-L82

**Comment by matthewp** [comment]
At this point we already have the dependency graph, it's in `data.graph`. Each key in the graph is the name of a module in the graph.  What I think we should do instead is try and match the bundles against items in that graph.  This would avoid going to the filesystem at all. 

What I really want to avoid is having any logic that checks for `~/` or the package name or any of that sort of stuff. steal-tools shouldn't have to know about `~/` at all, only the npm plugin should need to know about that.

That's why I was thinking that if you could turn a value like `app/components/**/*` and normalize that which becomes `app@1.0.0#components/**/*` you could then try to match each entry in the dependency graph which would find `app@1.0.0#components/tabs/tabs` for example. Make sense?

I know that we were normalizing it in that other function before, but I think that was a bad way to do it.

**Comment by pYr0x** [comment]
Thanks for your answer.
You say that we have a dependency graph at this point, but a bundle entry point needn't be in the dependency graph. 
http://forums.donejs.com/t/whats-the-different-between-sidebundle-and-bundle/206

**Comment by matthewp** [comment]
The dependency graph is everything that Steal loads. Nothing outside of the dependency graph gets written into a bundle (or a sideBundle) so we don't need to look at the filesystem, everything has already been fetched as this point.

**Comment by pYr0x** [comment]
```
if(window.location.hash === "#login") {
      System.import("site/login").then(function(){
        $("#main").login();
      });
    } else if(window.location.hash === "#signup" ) {
      System.import("site/signup").then(function(){
        $("#main").signup();
      });
    } else {
      System.import("site/homepage").then(function(){
        $("#main").homepage();
      });
    }
```
E.g. Site/login will not be in dependeny graph, right?

**Comment by matthewp** [comment]
Ugh, you're right, I was thinking this was only used after the full dependency graph was loaded. Arg, that not great.

In that case, I would rather just not accept the package name or `~/` to be included when using the `**/*` form. Let's throw an error if you try and use those. Either put real module names (like `app/components/foo` or use globs that go through the file system `components/**/*`. 

In this case I think we can remove the npm-specific like except perhaps the directories.lib part.

**Comment by matthewp** [comment]
@pYr0x Is this something you're still interested in working on? No pressure, of course, just wondering if it's more work than its worth...

**Comment by pYr0x** [comment]
i would like to finish it. but if have no idea when to do it ;)

**Comment by matthewp** [comment]
No problem. 🤠

**Comment by matthewp** [comment]
I'm going to close the PR for now. Feel free to reopen when it's ready.

**Pull request closed without merging**